"Failed to add user '%s' to realm '%s': user with username exists" Error displayed in logs if any user already exists in master realm  #16961

sjweixel opened the pull request:
Add user specified by environment variables KEYCLOAK_ADMIN and KEYCLOAK_ADMIN_PASSWORD.  Show different error if user exists with or without admin role.

Closes #16961

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

sjweixel committed:
Add user specified by environment variables KEYCLOAK_ADMIN and KEYCLOAK_ADMIN_PASSWORD.  Show different error if user exists with or with out admin role.

Closes keycloak#15789

mabartos reviewed (changes requested):
@sjweixel Thanks for the PR! Overall, looks good to me; I've just added some comments below. It's also necessary to create some tests to verify the proper solution.

mabartos commented:
@sjweixel You can inspire with writing tests here: https://github.com/sjweixel/keycloak/pull/1

mabartos commented:
Fixes #16961

vmuzikar commented on services/src/main/java/org/keycloak/services/ServicesLogger.java:
```diff
@@ -467,4 +467,9 @@ public interface ServicesLogger extends BasicLogger {
+
+    @LogMessage(level = ERROR)
+    @Message(id=107, value="User '%s' in realm '%s' is not assigned %s role")
+    void userNotAssignedRole(String user, String realm, String role);
```
@sjweixel Thanks for the contribution!

I wonder why we need this message. The error is that the user exists, not that it is not assigned an admin role.

I would maybe suggest a different approach for this PR. We could create a separate log message as suggested by @mabartos (something like `addAdminUserFailedAdminExists` maybe) and use it just in `KeycloakMain` without modifying `ApplienceBootstrap`.

vmuzikar replied:
Thanks for updating the PR! I believe the comment above is still valid.

sjweixel replied:
I'm not clear on the comment.  The code in ApplianceBootStrap is testing if the user exists and if so, determines if the admin role is assigned to the user.

```
        if (existingUser != null) {
            RoleModel adminRole = realm.getRole(AdminRoles.ADMIN);
            if (existingUser.hasRole(adminRole)) {
               ServicesLogger.LOGGER.addAdminUserUserExists(username, Config.getAdminRealm());
            } else {
               ServicesLogger.LOGGER.userNotAssignedRole(username, Config.getAdminRealm(), AdminRoles.ADMIN);
            }
            return;
        }
```
Is your comment addressing one of these concerns?
- The log message about the role doesn't match the logic in the code.
- The change doesn't match the issue as it is described.
- This message has no value.

The initial issue that I opened was #16961.  It was closed as a duplicate of this issue, then reopened after my initial PR.  One of my concerns is that when KEYCLOAK_ADMIN is used to specify a user, an admin user isn't created if that username already exists without admin role.  This may not be a situation that occurs often.   I think it could the user some headache if the situation arises.

sjweixel replied:
@vmuzikar.  Please explain your comment about changing code in KeycloakMain instead of ApplianceBootstrap.  I'm not certhan that what you are suggesting is possible.  The code that currently checks for existing users and generates error log is in ApplianceBootstrap.  If user exists check and log message is added to KeycloakMain, then ApplianceBootrap will need also need to be modified to remove existing user check and log message.

pedroigor commented:
Also not sure why we want an `INFO` for this one or any other message other than for a real error. Shall it be `DEBUG`?

mabartos committed:
Add tests for admin creation (#1)

sjweixel committed:
Removed unused import.
Add addAdminUserUserExists logger function.
Revert addUserFailedUserExists.
Updated test.

mabartos commented:
> Also not sure why we want an INFO for this one or any other message other than for a real error. Shall it be DEBUG?

Yep, it'd probably be better.

mabartos reviewed (commented):
@sjweixel Thanks for your updates, and sorry for the late response. 

I think this PR does more than it should (please see comment below). I've prepared separated PR for fixing the origination issue: https://github.com/keycloak/keycloak/pull/19922

If you still think the Keycloak behavior related to the admin creation lacks some behavior, please, create an issue for that and we can discuss the use cases more.

@vmuzikar WDYT?

sjweixel commented on services/src/main/java/org/keycloak/services/managers/ApplianceBootstrap.java:
```diff
@@ -95,9 +95,15 @@ public boolean createMasterRealm() {
         RealmModel realm = session.realms().getRealmByName(Config.getAdminRealm());
         session.getContext().setRealm(realm);
-
-        if (session.users().getUsersCount(realm) > 0) {
```
My PR was indented for issue #16961, which was closed as a duplicate of #15789 at the time the PR was created.  Issue #16961 has since been reopened and I will move my PR to that issue.  ENV vars should create an admin specified even if another admin exists.  It is the only way to recover if new admin deletes the initial admin account/revokes the initial user's admin role or if the initial admin username/password is somehow lost.   In older versions of Keycloak, the ENV would create a new admin user regardless of the existence of other admin users.   This was extremely helpful because it also made Keycloak recovery from backup database possible.  If a Keycloak instance is wiped out, but a database backup exists, you can recover without knowing initial admin credentials, because a new admin account could be created with ENVs.  I strongly believe there needs to be a way to create an admin account in the event that someone loses initial admin password.

vmuzikar commented:
@sjweixel Thank you for providing additional context. I believe I now understand the issue you are trying to fix here.

As you are correctly pointing out, we currently do not have an option to recover the admin account. This is currently being tracked by https://github.com/keycloak/keycloak/issues/19913. And as already mentioned by @mabartos, I'd also suggest to join the broader discussion around bootstrapping and recovering the initial admin user: https://github.com/keycloak/keycloak/discussions/8549.

The env vars for creating the initial admin are not meant to be used for account recovery. Otherwise it might lead to unintended side effects (accidentally re-creating the admin user which might be deleted on purpose, or accidentally resetting its credentials) by just not unsetting the env vars, which is standard e.g. in container environment. This would even create a potential security vulnerability.

Therefore the current approach seems valid to me:
https://github.com/keycloak/keycloak/blob/d7d50634b363348eadc63b9722cce5a7a0123616/services/src/main/java/org/keycloak/services/managers/ApplianceBootstrap.java#L99
And it is also aligned with what we do on the Welcome Screen:
https://github.com/keycloak/keycloak/blob/d7d50634b363348eadc63b9722cce5a7a0123616/services/src/main/java/org/keycloak/services/managers/ApplianceBootstrap.java#L55

I am going to close this PR and the [related issue](https://github.com/keycloak/keycloak/issues/16961). But I would like to ask you to join the aforementioned discussions. Thanks for understanding.
